src/mongo_odm/nvp.hpp

```diff
@@ -21,51 +21,76 @@
-            ar);                                                                                \
-    }                                                                                           \
-    template <class Archive, size_t I, size_t N>                                                \
-    typename std::enable_if<(I < N), void>::type mongo_odm_serialize_recur(Archive& ar) {       \
-        auto nvp = std::get<I>(mongo_odm_mapped_fields());                                      \
-        ar(cereal::make_nvp(nvp.name, this->*(nvp.t)));                                         \
-        mongo_odm_serialize_recur<Archive, I + 1, N>(ar);                                       \
-    }                                                                                           \
-                                                                                                \
-    template <class Archive, size_t I, size_t N>                                                \
-    typename std::enable_if<(I == N), void>::type mongo_odm_serialize_recur(Archive&) {         \
-        ;                                                                                       \
-    }
-
-// Register members and create serialize() function
-#define MONGO_ODM_MAKE_KEYS(Base, ...)                \
-    using mongo_odm_wrap_base = Base;                 \
-    constexpr static auto mongo_odm_mapped_fields() { \
-        return std::make_tuple(__VA_ARGS__);          \
-    }                                                 \
-    MONGO_ODM_SERIALIZE_KEYS
-
-// If using the mongo_odm::model, then also register _id as a field.
-#define MONGO_ODM_MAKE_KEYS_MODEL(Base, ...) \
-    MONGO_ODM_MAKE_KEYS(Base, MONGO_ODM_NVP(_id), __VA_ARGS__)
-
-#define MONGO_ODM_KEY(value) mongo_odm::hasCallIfFieldIsPresent<decltype(&value), &value>::call()
+#include <bsoncxx/types.hpp>
+
+#include <mongo_odm/macros.hpp>
+#include <mongo_odm/util.hpp>
 
 namespace mongo_odm {
 MONGO_ODM_INLINE_NAMESPACE_BEGIN
 
+template <typename NvpT, typename T>
+class NvpCRTP;
+
+template <typename NvpT>
+class SortExpr;
+
+// Forward declarations for Expression Types
+template <typename NvpT, typename U>
+class ComparisonExpr;
+
+template <typename NvpT, typename U>
+class ComparisonValueExpr;
+
+template <typename NvpT>
+class ModExpr;
+
+template <typename NvpT, typename U>
+class UpdateExpr;
+
+template <typename NvpT, typename U>
+class UpdateValueExpr;
+
+template <typename NvpT>
+class UnsetExpr;
+
+template <typename NvpT>
+class CurrentDateExpr;
+
+template <typename NvpT, typename U>
+class AddToSetUpdateExpr;
+
+template <typename NvpT, typename U, typename Sort = int>
+class PushUpdateExpr;
+
+struct current_date_t {
+    constexpr current_date_t() {
+    }
+};
+constexpr current_date_t current_date;
+
+// Forward declarations for Expression type trait structs
+
+template <typename T>
+struct is_sort_expression;
```

Comment:
Can you move all of these type traits to a consolidated header file and namespace? Either `impl` or `details`.